packages/react-app/src/components/giveaways/ParticipationButton.tsx

```diff
@@ -0,0 +1,144 @@
+import MuiAlert from '@mui/material/Alert';
+
+import useUserInfo from '../../hooks/useUserInfo';
+import { Giveaway, ParticipationState } from '../../lib/types';
+import FrontendApiClient from '../../services/backend';
+import ParticipationService from '../../services/giveawayparticipation';
+import { ActionButtonComponents } from './ActionButtons';
+import {
+  PendingLocationModal,
+  RejectionModal,
+  WinnerModal,
+} from './StatusModals';
+
+type ParticipationButtonProps = {
+  giveaway: Giveaway;
+  onStateChange?: (newState: ParticipationState) => void;
+};
+
+const ParticipationButton = ({
+  giveaway,
+  onStateChange,
+}: ParticipationButtonProps) => {
+  const [participationState, setParticipationState] =
+    useState<ParticipationState>(ParticipationState.CHECKING);
+
+  const userInfo = useUserInfo();
+  const [error, setError] = useState(false);
+  const [showPendingModal, setShowPendingModal] = useState(false);
+  const [showRejectedModal, setShowRejectedModal] = useState(false);
+  const [showWinnerModal, setShowWinnerModal] = useState(false);
+
+  const notifyRejection = useCallback(async () => {
+    if (!userInfo) return;
+    await FrontendApiClient.setNotifiedParticipant(
+      giveaway._id,
+      userInfo.email
+    );
+    setShowRejectedModal(true);
+  }, [giveaway, userInfo]);
+
+  const notifyWinner = useCallback(async () => {
+    if (!userInfo) return;
+    await FrontendApiClient.setNotifiedParticipant(
+      giveaway._id,
+      userInfo.email
+    );
+    setShowWinnerModal(true);
+  }, [giveaway, userInfo]);
+
+  const updateParticipationState = useCallback(
+    (fromEvent = false) => {
+      ParticipationService.getParticipationState(
+        giveaway,
+        undefined,
+        userInfo
+      ).then((state) => {
+        setParticipationState(state);
+        onStateChange?.(state);
+
+        if (fromEvent) {
+          if (state === ParticipationState.PENDING) {
+            setShowPendingModal(true);
+          }
+        }
+
+        if (state === ParticipationState.REJECTED) {
+          notifyRejection();
+        }
+
+        ParticipationService.shouldNotifyWinner(giveaway, userInfo).then(
+          (notify) => {
+            if (notify) {
+              notifyWinner();
+            }
+          }
+        );
+      });
+    },
+    [giveaway, userInfo, notifyRejection, notifyWinner, onStateChange]
+  );
+
+  const ActionButton: React.ReactNode = useMemo(() => {
+    let props = {};
+
+    if (participationState === ParticipationState.ALLOWED) {
+      props = {
+        giveaway: giveaway,
+        userInfo: userInfo,
+        successCallback: () => {
+          updateParticipationState(true);
+        },
+        errorCallback: () => {
+          setError(true);
+          updateParticipationState(true);
+        },
+      };
+    } else if (participationState === ParticipationState.MANAGE) {
+      props = { giveaway: giveaway };
+    }
+
+    return React.createElement(
```

Comment:
This ActionButton should be a functional components defined outside of this one like the rest of the components already created in the app